db/memtable.cc

```diff
@@ -842,6 +846,66 @@ bool MemTable::Get(const LookupKey& key, std::string* value, Status* s,
   return found_final_value;
 }
 
+bool MemTable::MultiGet(const ReadOptions& read_options, MultiGetRange* range,
+                        ReadCallback* callback, bool* is_blob) {
+  // The sequence number is updated synchronously in version_set.h
+  if (IsEmpty()) {
+    // Avoiding recording stats for speed.
+    return false;
+  }
+  std::bitset<MultiGetContext::MAX_BATCH_SIZE> keys;
+  int idx = 0;
+  for (auto iter = range->begin(); iter != range->end(); ++iter) {
+    Slice user_key = iter->lkey->user_key();
+    size_t ts_sz =
+        GetInternalKeyComparator().user_comparator()->timestamp_size();
+    bool may_contain = true;
+    if (bloom_filter_) {
+      // when both memtable_whole_key_filtering and prefix_extractor_ are set,
+      // only do whole key filtering for Get() to save CPU
+      if (moptions_.memtable_whole_key_filtering) {
+        may_contain = bloom_filter_->MayContain(
+            StripTimestampFromUserKey(user_key, ts_sz));
+      } else {
+        assert(prefix_extractor_);
+        may_contain =
+            !prefix_extractor_->InDomain(user_key) ||
+            bloom_filter_->MayContain(prefix_extractor_->Transform(user_key));
+      }
+
+      if (may_contain) {
+        keys.set(idx);
```

Comment:
If ```DynamicBloom::MayContain()``` is passed a MultiGetRange, it can just call ```SkipKey(iter)``` on that range to skip a key that got a negative result in the bloom filter lookup. Then all you have to do later is just iterate through that range to find keys that may be present in the memtable. There'd be no need for the ```keys``` bitset.